tornado/gen.py

```diff
@@ -185,6 +185,45 @@ def coroutine(func):
     From the caller's perspective, ``@gen.coroutine`` is similar to
     the combination of ``@return_future`` and ``@gen.engine``.
     """
+    return _make_coroutine_wrapper(func)
+
+def cython_coroutine(func, replace_callback=True):
+    """Decorator for asynchronous generators.
+
+    Any generator that yields objects from this module must be wrapped
+    in either this decorator or `engine`.
+
+    Coroutines may "return" by raising the special exception
+    `Return(value) <Return>`.  In Python 3.3+, it is also possible for
+    the function to simply use the ``return value`` statement (prior to
+    Python 3.3 generators were not allowed to also return values).
+    In all versions of Python a coroutine that simply wishes to exit
+    early may use the ``return`` statement without a value.
+
+    Functions with this decorator return a `.Future`.  Additionally,
+    they may be called with a ``callback`` keyword argument, which
+    will be invoked with the future's result when it resolves.  If the
+    coroutine fails, the callback will not be run and an exception
+    will be raised into the surrounding `.StackContext`.  The
+    ``callback`` argument is not visible inside the decorated
+    function; it is handled by the decorator itself.
+
+    From the caller's perspective, ``@gen.coroutine`` is similar to
+    the combination of ``@return_future`` and ``@gen.engine``.
+
+    This version works with cythonized coroutines by assuming the
+    decorated function is a generator (instead of checking).
+    Use reuglar @gen.coroutine unless necessary.
+    """
+    return _make_coroutine_wrapper(func, assume_generator=True)
```

Comment:
I choose the branch 3.1 as the base branch. But it seems that the branch doesn't has the commit. I am not sure on which branch it was done